source/extensions/filters/http/api_key_auth/api_key_auth.h

```diff
@@ -0,0 +1,189 @@
+#pragma once
+
+#include "envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.pb.h"
+#include "envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.pb.validate.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "source/common/common/logger.h"
+#include "source/extensions/filters/http/common/pass_through_filter.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ApiKeyAuth {
+
+/**
+ * All Basic Auth filter stats. @see stats_macros.h
+ */
+#define ALL_API_KEY_AUTH_STATS(COUNTER)                                                            \
+  COUNTER(allowed)                                                                                 \
+  COUNTER(unauthorized)                                                                            \
+  COUNTER(forbidden)
+
+/**
+ * Struct definition for API key auth stats. @see stats_macros.h
+ */
+struct ApiKeyAuthStats {
+  ALL_API_KEY_AUTH_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+using ApiKeyAuthProto = envoy::extensions::filters::http::api_key_auth::v3::ApiKeyAuth;
+using ApiKeyAuthPerRouteProto =
+    envoy::extensions::filters::http::api_key_auth::v3::ApiKeyAuthPerRoute;
+using KeySourcesProto = envoy::extensions::filters::http::api_key_auth::v3::KeySources;
+
+// Credentials is a map of API key to client ID.
+using Credentials = absl::flat_hash_map<std::string, std::string>;
+
+/**
+ * The sources to get the API key from the incoming request.
+ */
+class KeySources {
+public:
+  KeySources(const KeySourcesProto& proto_config);
+
+  /**
+   * To get the API key from the incoming request.
+   * @param headers the incoming request headers.
+   * @param key_buffer the buffer to store the API key.
+   * @return the result of getting the API key.
+   */
+  absl::string_view getKey(const Http::RequestHeaderMap& headers, std::string& buffer) const;
```

Comment:
The API is not clean: the "buffer" is more of a performance enhancement and it is also passed as the return value.
I guess this could've been avoided if polymorphism/inheritance was used, but I'm not a c++ expert that has a strong opinion here.

Reply:
Yeah, it's used for performance. When the header key is used, we can simpliy return a string_view. If query key or cookie key is used, we need create a new string. 🤔

Reply:
Yeah, I understand that. While this is not optimal (why is the string being allocated/passed if it is not used in some case), I think this can be refactored in the future if needed.

Ac couple of idea (more work, not part of this PR):
1. Changing the API to use something else than "getters" -> use an action instead (e.g., `validateKey()`).
2. Creating a new type that is based on `std::variant<std::string, std::string_view>` but implements the string API (I haven't thoroughly hashed this out, and there may be gaps in my thinking).